tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -212,7 +218,383 @@
+    } else {
+      // TODO handle Ragged Tensors????
+      // [y_pred,
+      //    y_true], _ = ragged_assert_compatible_and_get_flat_values([y_pred, y_true],
+      //                                                   sampleWeights)
+      numThresholds = tf.constant(thresholds.length);
+      oneThresh = tf.constant(true);
+    }
+
+    List<Op> controlOps = new ArrayList<>();
+    Operand<TInt32> axes = allAxes(tf, lPredictions);
+    controlOps.add(
+        tf.withSubScope("updateConfusionMatrixVariables-1")
+            .assertThat(
+                tf.reduceAll(
+                    tf.math.greaterEqual(
+                        lPredictions, cast(tf, tf.constant(0), lPredictions.type())),
+                    axes),
+                Collections.singletonList(tf.constant("predictions must be >= 0"))));
+    controlOps.add(
+        tf.withSubScope("updateConfusionMatrixVariables-2")
+            .assertThat(
+                tf.reduceAll(
+                    tf.math.lessEqual(lPredictions, cast(tf, tf.constant(1), lPredictions.type())),
+                    axes),
+                Collections.singletonList(tf.constant("predictions must be <= 1"))));
+
+    LossTuple<T> result =
+        LossesHelper.squeezeOrExpandDimensions(tf, lLabels, lPredictions, lSampleWeight);
+    lPredictions = result.getTarget();
+    lLabels = result.getLabels();
+    lSampleWeight = result.getSampleWeights();
+
+    if (!lPredictions.shape().isCompatibleWith(lLabels.shape()))
+      throw new IllegalArgumentException(
+          String.format(
+              "Shapes %s and %s are incompatible)",
+              lPredictions.shape().toString(), lLabels.asOutput().shape().toString()));
+
+    if (topK != null) {
+      lPredictions = filterTopK(tf, lPredictions, topK);
+    }
+
+    if (classId != null) {
+      lLabels = tf.squeeze(tf.gather(lLabels, tf.constant(new int[] {classId}), tf.constant(1)));
+      lPredictions =
+          tf.squeeze(tf.gather(lPredictions, tf.constant(new int[] {classId}), tf.constant(1)));
+      lLabels = tf.expandDims(lLabels, tf.constant(0));
+      lPredictions = tf.expandDims(lPredictions, tf.constant(0));
+    }
+    org.tensorflow.op.core.Shape<TInt32> predShape = tf.shape(lPredictions);
+    Operand<TInt32> numPredictions =
+        tf.reshape(tf.shape.size(lPredictions, tf.constant(0)), tf.constant(Shape.scalar()));
+    Operand<TInt32> numLabels =
+        tf.select(
+            tf.math.equal(tf.shape.numDimensions(predShape), tf.constant(1)),
+            tf.constant(1),
+            tf.reduceProd(
+                tf.shape.takeLast(
+                    predShape, tf.math.sub(tf.shape.numDimensions(predShape), tf.constant(1))),
+                tf.constant(0)));
+    Operand<TInt32> threshLabelTile = tf.select(oneThresh, numLabels, tf.constant(1));
+
+    Operand<T> predictionsExtraDim;
+    Operand<TBool> labelsExtraDim;
+    if (multiLabel) {
```

Comment:
It would be really helpful to document the shapes and semantics of the `...ExtraDim` variables. By the time we get here, we started with prediction and label vectors that had both flexible and case-dependent shapes, and we've reshaped them multiple times.

Reply:
The ExtraDims are added so the operands of the tile operations later on are compatible. I will add a comment.